src/Microsoft.ML.FastTree/FastTreeArguments.cs

```diff
@@ -20,19 +20,81 @@ internal interface IFastTreeTrainerFactory : IComponentFactory<ITrainer>
     {
     }
 
+    /// <summary>
+    /// Stopping measurements for classification and regression.
+    /// </summary>
+    public enum EarlyStoppingMetric
+    {
+        /// <summary>
+        /// L1-norm of gradient.
+        /// </summary>
+        L1Norm = 1,
+        /// <summary>
+        /// L2-norm of gradient.
+        /// </summary>
+        L2Norm = 2
+    };
+
+    /// <summary>
+    /// Stopping measurements for ranking.
+    /// </summary>
+    public enum EarlyStoppingRankingMetric
+    {
+        /// <summary>
+        /// NDCG@1
+        /// </summary>
+        NdcgAt1 = 1,
+        /// <summary>
+        /// NDCG@3
+        /// </summary>
+        NdcgAt3 = 3
+    }
+
     /// <include file='doc.xml' path='doc/members/member[@name="FastTree"]/*' />
     public sealed partial class FastTreeBinaryClassificationTrainer
     {
         [TlcModule.Component(Name = LoadNameValue, FriendlyName = UserNameValue, Desc = Summary)]
         public sealed class Options : BoostedTreeOptions, IFastTreeTrainerFactory
         {
+
             /// <summary>
             /// Option for using derivatives optimized for unbalanced sets.
             /// </summary>
             [Argument(ArgumentType.LastOccurenceWins, HelpText = "Option for using derivatives optimized for unbalanced sets", ShortName = "us")]
             [TGUI(Label = "Optimize for unbalanced")]
             public bool UnbalancedSets = false;
 
+            /// <summary>
+            /// internal state of <see cref="EarlyStoppingMetric"/>. It should be always synced with
+            /// <see cref="BoostedTreeOptions.EarlyStoppingMetrics"/>.
+            /// </summary>
+            // Disable 649 because Visual Studio can't detect its assignment via property.
+            #pragma warning disable 649
+            private EarlyStoppingMetric _earlyStoppingMetric;
+            #pragma warning restore 649
+
+            /// <summary>
+            /// Early stopping metrics.
+            /// </summary>
+            public EarlyStoppingMetric EarlyStoppingMetric
```

Comment:
>EarlyStoppingMetric [](start = 19, length = 19)

From comments for previous option it looks like it should be applicable only for regression and ranking problem. Does it actually works for binary classification?

Reply:
I don't exactly know but conceptually yes. It can be used to measure gradient norm.

---
In reply to: [262295829](https://github.com/dotnet/machinelearning/pull/2753#discussion_r262295829) [](ancestors = 262295829)